apps/music/js/queue.js

```diff
@@ -0,0 +1,335 @@
+/* global asyncStorage, Database */
+/* exported PlaybackQueue */
+'use strict';
+
+var PlaybackQueue = (function() {
+
+  // The key for storing the repeat and shuffle options.
+  var SETTINGS_OPTION_KEY = 'settings_option_key';
+
+  var Repeat = {
+    OFF: 0,
+    LIST: 1,
+    SONG: 2,
+
+    /**
+     * Return the "next" repeat setting in its canonical order; OFF -> LIST ->
+     * SONG -> OFF.
+     *
+     * @param {Repeat} val The current value.
+     * @return {Repeat} The next value.
+     */
+    next: function(val) {
+      return (val + 1) % 3;
+    }
+  };
+
+  /**
+   * Return a % b, ensuring that the result is always positive.
+   *
+   * @param {Number} a The dividend.
+   * @param {Number} b The divisor.
+   * @return {Number} The positive remainder.
+   */
+  function posmod(a, b) {
+    var r = a % b;
+    return r < 0 ? r + b : r;
+  }
+
+  /**
+   * Fill an array with values from 0..n.
+   *
+   * @param {Number} length The length of the array.
+   * @return {Array} The array.
+   */
+  function fillIndices(length) {
+    var indices = new Array(length);
+    for (var i = 0; i < indices.length; i++) {
```

Comment:
micro-optimization: change `indices.length` to `length` (avoids having the array fetch its length on every iteration of the loop)